client/allocrunner/taskrunner/stats_hook.go

```diff
@@ -48,13 +49,14 @@ func (h *statsHook) Poststart(ctx context.Context, req *interfaces.TaskPoststart
 	defer h.mu.Unlock()
 
 	// This shouldn't happen, but better safe than risk leaking a goroutine
-	if h.stopCh != nil {
+	if h.cancel != nil {
 		h.logger.Debug("poststart called twice without exiting between")
-		close(h.stopCh)
+		h.cancel()
 	}
 
-	h.stopCh = make(chan struct{})
-	go h.collectResourceUsageStats(req.DriverStats, h.stopCh)
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Using the passed-in context means this context will get canceled when the task is killed.

That seems fine -- even optimal probably -- but it's such a subtle change I wanted to call it out in case anybody can think of a reason it shouldn't be this way. :)

Reply:
So before stats was signaled to stop when the Exited hook ran, rather than when the task was killed. Hmm then nilling out the cancel is probably over kill then? Can we just pass the context all together without a child cancel and get rid of the Exited hook?

Reply:
Ugh, this is a good question, and I believe the answer has changed over time.

Initially we only used contexts for hooks like a request/response: useful for cancelling long running hooks, but *not* for signalling task state changes.

Now we can signal task state changes for hooks with background goroutines through a chan. This is convenient, but it's an emergent property of the API, not an intentional design choice.

I _think_ we ideally want to treat contexts as being request scoped (where "request" == a hook method), so hooks operate like external plugins (as we might want to make them plugins in the future).

If we want to enforce that we should wrap Poststart's context is a subcontext that is cancelled after each call to avoid developers using it instead of Exited.